// @oracle/oraclejet-preact/es/UNSAFE_Radio.js
/* @oracle/oraclejet-preact: 15.1.5 */
import { jsx, jsxs } from 'preact/jsx-runtime';
import { useComponentTheme } from './hooks/UNSAFE_useComponentTheme.js';
import { useFocus } from './hooks/UNSAFE_useFocus.js';
import { useFormContext } from './hooks/UNSAFE_useFormContext.js';
import { useFormFieldContext } from './hooks/UNSAFE_useFormFieldContext.js';
import { useId } from './hooks/UNSAFE_useId.js';
import { useTabbableMode } from './hooks/UNSAFE_useTabbableMode.js';
import { useInteractionStyle } from './hooks/UNSAFE_useInteractionStyle.js';
import { u as useCheckboxRadioContext } from './CheckboxRadioContext-172f4c19.js';
import { CheckIcon as SvgCheck } from './PRIVATE_ThemedIcons/CheckIcon.js';
import { RadioOffIcon as SvgRadioButtonOff } from './PRIVATE_ThemedIcons/RadioOffIcon.js';
import { RadioOnIcon as SvgRadioButtonOn } from './PRIVATE_ThemedIcons/RadioOnIcon.js';
import './UNSAFE_HiddenAccessible.js';
import { u as useRadioGroupContext } from './RadioGroupContext-762f0e44.js';
import { IconUserAssistance } from './UNSAFE_UserAssistance.js';
import './utils/UNSAFE_stringUtils.js';
import { radioIconVars } from './UNSAFE_Radio/themes/RadioIconContract.css.js';
import { RadioIconRedwoodTheme } from './UNSAFE_Radio/themes/redwood/RadioIconTheme.js';
import { RadioRedwoodTheme } from './UNSAFE_Radio/themes/redwood/RadioTheme.js';
import { H as HiddenAccessible } from './HiddenAccessible-18327856.js';
import { g as getVarName } from './stringUtils-0805904b.js';
import './utils/UNSAFE_logger.js';
import './_curry1-4637aee9.js';
import './_curry2-da491c7e.js';
import './hooks/UNSAFE_useToggle.js';
import 'preact/hooks';
import 'preact';
import './utils/UNSAFE_mergeProps.js';
import './utils/PRIVATE_clientHints.js';
import './clientHints-6ff785cb.js';
import './hooks/UNSAFE_useHover.js';
import './hooks/UNSAFE_useActive.js';
import './UNSAFE_Icon.js';
import './Icon-44f1fb0b.js';
import './utils/UNSAFE_size.js';
import './UNSAFE_Theme.js';
import './utils-79fd9e1a.js';
import './Common/themes/themeContract.css.js';
import './hooks/UNSAFE_useTooltip.js';
import './UNSAFE_Floating.js';
import './Floating-44024b63.js';
import 'preact/compat';
import './index-a9b88c2d.js';
import './hooks/UNSAFE_useUser.js';
import './index-ce660c32.js';
import './Common.js';
import './Common/themes.js';
import './Common/themes/redwood/theme.js';
import './utils/PRIVATE_floatingUtils.js';
import './utils/PRIVATE_refUtils.js';
import './hooks/UNSAFE_useOutsideClick.js';
import './utils/UNSAFE_arrayUtils.js';
import './utils/UNSAFE_classNames.js';
import './classNames-6df97f98.js';
import './UNSAFE_Floating/themes/redwood/FloatingTheme.js';
import './UNSAFE_Floating/themes/FloatingStyles.css.js';
import './FloatingStyles.styles.css';
import './UNSAFE_Floating/themes/redwood/FloatingBaseTheme.css.js';
import './UNSAFE_Floating/themes/redwood/FloatingBaseTheme.styles.css';
import './UNSAFE_Floating/themes/redwood/FloatingVariants.css.js';
import './vanilla-extract-recipes-createRuntimeFn.esm-ea2e9956.js';
import './vanilla-extract-dynamic.esm-22241307.js';
import './UNSAFE_Floating/themes/FloatingContract.css.js';
import './UNSAFE_Layer.js';
import './useThemeInterpolations-18331a05.js';
import './hooks/UNSAFE_useColorScheme.js';
import './hooks/UNSAFE_useScale.js';
import './utils/UNSAFE_interpolations/theme.js';
import './utils/UNSAFE_mergeInterpolations.js';
import './_curry3-7db59ba0.js';
import './_has-92b39746.js';
import './hooks/UNSAFE_useTouch.js';
import './hooks/UNSAFE_useAnimation.js';
import './useAnimation-6ccd68bf.js';
import './hooks/UNSAFE_useThemeInterpolations.js';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentTheme.js';
import './hooks/UNSAFE_useTooltip/themes/TooltipContentStyles.css.js';
import './TooltipContentStyles.styles.css';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentBaseTheme.css.js';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentBaseTheme.styles.css';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentVariants.css.js';
import './TooltipContentVariants.styles2.css';
import './UNSAFE_Icon/themes/IconStyle.css.js';
import './IconStyle.styles.css';

import './InlineHelpSource-d3bbf3f1.js';
import './hooks/UNSAFE_useTranslationBundle.js';
import './UNSAFE_UserAssistance/themes/redwood/UserAssistanceTheme.js';
import './UNSAFE_UserAssistance/themes/UserAssistanceStyles.css.js';
import './UserAssistanceStyles.styles.css';
import './UNSAFE_UserAssistance/themes/redwood/UserAssistanceBaseTheme.css.js';
import './UNSAFE_UserAssistance/themes/redwood/UserAssistanceBaseTheme.styles.css';
import './UNSAFE_UserAssistance/themes/redwood/UserAssistanceVariants.css.js';
import './UNSAFE_ComponentMessage.js';
import './ComponentMessage-f1feb94a.js';
import './PRIVATE_Message.js';
import './UNSAFE_Flex.js';
import './Flex-1ec75561.js';
import './utils/UNSAFE_interpolations/dimensions.js';
import './utils/UNSAFE_interpolations/boxalignment.js';
import './keys-6a8847c4.js';
import './utils/UNSAFE_interpolations/flexbox.js';
import './flexbox-94918d4c.js';
import './utils/UNSAFE_interpolations/flexitem.js';
import './flexitem-cf0b1a33.js';
import './utils/PRIVATE_timer.js';
import './PRIVATE_Message/themes/MessageStyles.css.js';
import './MessageStyles.styles.css';
import './MessageCloseButton-f4c6ebd4.js';
import './UNSAFE_IconButton.js';
import './IconButton-080dd7b5.js';
import './UNSAFE_BaseButton.js';
import './BaseButton-f4a900bd.js';
import './hooks/UNSAFE_usePress.js';
import './UNSAFE_ButtonLayout.js';
import './ButtonLayout-d77f4c52.js';
import './UNSAFE_Text.js';
import './Text-5d60dbd2.js';
import './UNSAFE_Text/themes/TextStyles.css.js';
import './TextStyles.styles.css';
import './PRIVATE_ThemedIcons/CloseIcon.js';
import './MessageDetail-f7fc7fa8.js';
import './MessageFormattingUtils-24e5f4ed.js';
import './utils/UNSAFE_getLocale.js';
import './Message.types-a7a0a35b.js';
import './MessageStartIcon-f459d4a6.js';
import './PRIVATE_ThemedIcons/MessageConfirmationIcon.js';
import './PRIVATE_ThemedIcons/MessageErrorIcon.js';
import './PRIVATE_ThemedIcons/MessageInfoIcon.js';
import './PRIVATE_ThemedIcons/MessageWarningIcon.js';
import './MessageSummary-ca8ee040.js';
import './MessageTimestamp-5727cf78.js';
import './MessageUtils-42e0fc95.js';
import './utils/UNSAFE_soundUtils.js';
import './UNSAFE_MessageBanner/themes/redwood/MessageBannerTheme.js';
import './UNSAFE_MessageBanner/themes/MessageBannerStyles.css.js';
import './MessageBannerStyles.styles.css';
import './UNSAFE_MessageBanner/themes/redwood/MessageBannerBaseTheme.css.js';
import './UNSAFE_MessageBanner/themes/redwood/MessageBannerBaseTheme.styles.css';
import './UNSAFE_MessageBanner/themes/redwood/MessageBannerVariants.css.js';
import './MessageBannerVariants.styles.css';
import './MessagesManager-a46ef486.js';
import './PRIVATE_TransitionGroup.js';
import './hooks/UNSAFE_useMessagesContext.js';
import './UNSAFE_ComponentMessage/themes/redwood/ComponentMessageTheme.js';
import './UNSAFE_ComponentMessage/themes/ComponentMessageStyles.css.js';
import './ComponentMessageStyles.styles.css';
import './UNSAFE_ComponentMessage/themes/redwood/ComponentMessageBaseTheme.css.js';
import './UNSAFE_ComponentMessage/themes/redwood/ComponentMessageBaseTheme.styles.css';
import './UNSAFE_ComponentMessage/themes/redwood/ComponentMessageVariants.css.js';
import './ComponentMessageContainer-5917bcb5.js';
import './hooks/UNSAFE_useInputGroupContext.js';
import './UNSAFE_Popup.js';
import './utils/PRIVATE_tabbableUtils.js';
import './_arity-249f1f1f.js';
import './UNSAFE_FocusTrap.js';
import './FocusTrap-46648548.js';
import './hooks/PRIVATE_useAnimationStatus.js';
import './UNSAFE_Popup/themes/PopupContract.css.js';
import './UNSAFE_Popup/themes/redwood/PopupTheme.js';
import './UNSAFE_Popup/themes/redwood/PopupBaseTheme.css.js';
import './UNSAFE_Popup/themes/redwood/PopupBaseTheme.styles.css';
import './UNSAFE_Popup/themes/redwood/PopupVariants.css.js';
import './PopupStyles.styles.css';
import './UNSAFE_Divider.js';
import './Divider-721a761c.js';
import './UNSAFE_Divider/themes/DividerStyles.css.js';
import './DividerStyles.styles.css';
import './CompactHelpSource-11fe262c.js';
import './PRIVATE_ThemedIcons/HelpIcon.js';
import './hooks/UNSAFE_useFocusWithin.js';
import './useFocusWithin-d572dd3b.js';
import './UNSAFE_Label/themes/redwood/LabelTheme.js';
import './UNSAFE_Label/themes/LabelStyles.css.js';
import './LabelStyles.styles.css';
import './UNSAFE_Label/themes/redwood/LabelBaseTheme.css.js';
import './UNSAFE_Label/themes/redwood/LabelBaseTheme.styles.css';
import './UNSAFE_Label/themes/redwood/LabelVariants.css.js';
import './UNSAFE_Radio/themes/RadioIconStyles.css.js';
import './RadioIconStyles.styles.css';
import './UNSAFE_Radio/themes/redwood/RadioIconBaseTheme.css.js';
import './UNSAFE_Radio/themes/redwood/RadioIconBaseTheme.styles.css';
import './UNSAFE_Radio/themes/redwood/RadioIconVariants.css.js';
import './RadioIconVariants.styles2.css';
import './UNSAFE_Radio/themes/RadioStyles.css.js';
import './RadioStyles.styles.css';
import './UNSAFE_Radio/themes/redwood/RadioBaseTheme.css.js';
import './UNSAFE_Radio/themes/redwood/RadioBaseTheme.styles.css';
import './UNSAFE_Radio/themes/redwood/RadioVariants.css.js';

const Radio = ({ value, children, ...assistanceProps }) => {
    const { isReadonly: isFormReadonly, isDisabled: isFormDisabled } = useFormContext();
    const { isDisabled: isFieldDisabled, isReadonly: isFieldReadonly } = useFormFieldContext();
    const { name, value: groupValue, onCommit } = useRadioGroupContext();
    const { isFocusRingShown } = useCheckboxRadioContext();
    const { isTabbable } = useTabbableMode();
    const id = useId();
    const { isFocus, focusProps } = useFocus();
    const isSelected = groupValue === value;
    const isDisabled = isFieldDisabled ?? isFormDisabled;
    const isReadonly = isFieldReadonly ?? isFormReadonly;
    const { interactionProps, applyActiveStyle } = useInteractionStyle();
    const { classes: radioClasses, styles: { valueLabelStyle, valueTextStyle } } = useComponentTheme(RadioRedwoodTheme, {
        disabled: isDisabled ? 'isDisabled' : 'notDisabled',
        focused: isFocus && isFocusRingShown ? 'isFocused' : ''
    });
    const { classes: radioIconClasses } = useComponentTheme(RadioIconRedwoodTheme, {
        disabled: isDisabled ? 'isDisabled' : 'notDisabled',
        selected: isSelected ? 'isSelected' : 'notSelected',
        readonly: isReadonly ? 'isReadonly' : 'notReadonly',
        active: applyActiveStyle ? 'isActive' : 'notActive'
    });
    const { assistiveText, helpSourceLink } = assistanceProps;
    const hasUserAssistive = !isReadonly && !!(assistiveText || helpSourceLink);
    const uaId = hasUserAssistive ? `${id}-icon-ua` : undefined;
    const IconComp = isReadonly
        ? isSelected
            ? SvgCheck
            : SvgRadioButtonOff
        : isSelected
            ? SvgRadioButtonOn
            : SvgRadioButtonOff;
    return (jsx("div", { children: jsxs("label", { class: radioClasses, ...interactionProps, children: [jsx(HiddenAccessible, { children: jsx("input", { type: "radio", name: name, id: id, value: value, ...(!isReadonly && {
                            onChange: (e) => {
                                e.stopPropagation();
                                onCommit?.({ previousValue: groupValue, value });
                            }
                        }), checked: isSelected, disabled: isDisabled, "aria-describedby": uaId, ...(!isTabbable && { tabIndex: -1 }), ...(isReadonly && { 'aria-readonly': true }), ...focusProps }) }), jsx("span", { "aria-hidden": "true", class: radioIconClasses, children: jsx(IconComp, { size: getVarName(radioIconVars.iconSize), color: "currentColor" }) }), jsxs("span", { class: valueLabelStyle, children: [jsx("span", { class: valueTextStyle, children: children }), hasUserAssistive && (jsx(IconUserAssistance, { ...assistanceProps, isTabbable: isTabbable && isSelected && isFocus, id: uaId }))] })] }) }));
};

export { Radio };
