// @oracle/oraclejet-preact/es/utils/UNSAFE_interpolations/boxalignment.js
/* @oracle/oraclejet-preact: 15.1.5 */
import "./boxalignment.css";
import { k as keys_1 } from '../../keys-6a8847c4.js';
import { sizeToCSS } from '../UNSAFE_size.js';
import { coerceArray } from '../UNSAFE_arrayUtils.js';
import '../../_curry1-4637aee9.js';
import '../../_has-92b39746.js';
import '../../UNSAFE_Theme.js';
import '../../utils-79fd9e1a.js';
import '../../Common/themes/themeContract.css.js';

// boxalignment are styles that help with alignment for Grid/Flex styles, like justifyContent,
// align, and gap.
const alignStyles = {
  baseline: "nj20ma",
  center: "on2s0t",
  end: "_1exn92c",
  start: "srs9c5",
  inherit: "_1cbft8y",
  initial: "_1s6qy7c",
  stretch: "_1ybhh0n"
};
const justifyStyles = {
  center: "b3ptzr",
  end: "zc68z7",
  start: "ejnawe",
  inherit: "_31srwp",
  initial: "_1v520qd",
  around: "_1483tu5",
  between: "_5onvdh",
  evenly: "u27i6d"
};
const boxAlignmentInterpolations = {
  align: ({
    align
  }) => align === undefined ? {} : {
    class: alignStyles[align]
  },
  justify: ({
    justify
  }) => justify === undefined ? {} : {
    class: justifyStyles[justify]
  },
  // See https://developer.mozilla.org/en-US/docs/Web/CSS/gap
  gap: ({
    gap
  }) => {
    if (gap === undefined) {
      return {};
    } else {
      const [rowSize, columnSize = rowSize] = coerceArray(gap);
      return {
        gap: `${sizeToCSS(rowSize)} ${sizeToCSS(columnSize)}`
      };
    }
  }
};
// These consts are exported so they can be used in stories to show the allowed options
// e.g., See Flex.stories.tsx:   align: {control: 'select',options: aligns}
const aligns = keys_1(alignStyles);
const justifies = keys_1(justifyStyles);

export { aligns, boxAlignmentInterpolations, justifies };
